### PR TITLE
Ported NUT package to PFSense 2.3 pkg structure.

### PR DESCRIPTION
**Pull request by sander1988**
NUT package for PFSense 2.3 . Fixed several bugs that where introduced due to PFSense 2.3 and converted the interface to bootstrap. This includes the widget.

The NUT package for FreeBSD was build with the default options during testing.

I have tested it as far as I can, but I only have one UPS available for testing. So more people should test it.

### CODE REVIEW

File: sysutils/pfSense-pkg-nut/BuildPackage.sh [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+echo "Creating new package..."
+make clean 
+make package 
+echo "Done"
+
+echo "Uploading new package to PFSense test server (10.0.201.150)"
+scp -P 2788 work/pkg/*.txz root@10.0.201.150:.
+echo "Done"
+
+exit 0
```
Comment (rbgarga) [inline sysutils/pfSense-pkg-nut/BuildPackage.sh]: Looks like you forgot your own test script and committed it together, can you please remove it?

File: sysutils/pfSense-pkg-nut/Makefile [hunk 2]
```diff
@@ -0,0 +1,62 @@
+# Created by: Sander Peterse <sander.peterse88@gmail.com>
+# $FreeBSD$
+
+PORTNAME=		pfSense-pkg-nut
+PORTVERSION=	2.3.0
+PORTREVISION=	1
```
Comment (rbgarga) [inline sysutils/pfSense-pkg-nut/Makefile]: New packages should not contain PORTREVISION line, please remove

**Comment by PiBa-NL** [comment]
Sadly it looks like at least one of us wasted some time.. https://github.com/pfsense/FreeBSD-ports/pull/111

**Comment by sander1988** [comment]
Yep, found out yesterday. I checked before but probably one ore two days before your merge request.

Maybe we should merge it and keep the best of both?

You can also contact me directly, email can be found in the source code header comments.

**Comment by sander1988** [comment]
I made the requested changes (commit d7208cb). In a few days I will push an updated version. After contact with @PiBa-NL we decided to merge his and mine version of the NUT package and keep the best of both.

**Comment by sbeaver-netgate** [comment]
So pleased you were able to get together on this. Each version had merit but together everyone will win.

**Comment by woffko** [comment]
Sorry for the stupid question, when it will be available for testing? If it is already available then what should I do to test it? Thanks.

**Comment by sander1988** [comment]
The package of PiBa-NL and mime have been merged, several bugs fixed and other small improvements. Commits fddc5ae and 64f3b26 .

Please test and send me your feedback.

File: sysutils/pfSense-pkg-nut/files/usr/local/pkg/nut.xml [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+	<fields>
+		<field>
+			<name>General Settings</name>
+			<type>listtopic</type>
+		</field>
+		<field>
+			<fielddescr>UPS Monitoring</fielddescr>
+			<fieldname>monitor</fieldname>
+			<type>select</type>
+			<required/>
```
Comment (PiBa-NL) [inline sysutils/pfSense-pkg-nut/files/usr/local/pkg/nut.xml]: Can you remove the 'required' ? Its not possible to disable nut with the "Disabled" option from line 86, which fails to save as it has no value.. (Or add another default empty option, with the choice to specifically disable nut?)

Comment (sander1988) [inline sysutils/pfSense-pkg-nut/files/usr/local/pkg/nut.xml]: Fixed in commit 781492c

**Comment by Andrew17856** [comment]
Hi Sander1988. Thanks for your work on this.  I'm very happy to test the package.  What is the easiest way to deploy it onto a live pfSense appliance please?

**Comment by sander1988** [comment]
The easiest way is building the pkg package, uploading it to your PFSense box and install it from the command line.

You can download my x64 builds from my drive (so you can skip building the pkg packages yourself): https://drive.google.com/open?id=0BwLY6Pu4pV3QdTNDeXRUVzRtNEE .

First you have to install nut:
# pkg add nut-2.7.4_x64.txz

And then you can install the PFSense package for nut:
# pkg add pfSense-pkg-nut-2.3.0_x64.txz

Note: do NOT install it on a production box.

**Comment by Andrew17856** [comment]
Thanks.   I've installed the packages, but can't get it to work with my UPS.

I suspect half the reason is I'm not sure what UPS to select from the menu (and in particular what, if anything to select in Local UPS Generic Type and Local UPS Cable Type).

I'll try playing further tomorrow.

During boot, my UPS is recognised as follows:

May 6 00:52:01 	kernel 		uhid0: <MEC MEC0002, class 0/0, rev 1.00/1.00, addr 1> on usbus0
May 6 00:52:01 	kernel 		ugen0.2: <MEC> at usbus0 

I've seen forum posts suggesting the MEC0003 protocol works well with the Blazer driver

However, I get various messages are in the nut log, e.g.:
May 6 01:23:11 	upsd 	34473 	Can't connect to UPS [pfSense] (blazer_usb-pfSense): No such file or directory 

May 6 01:23:11 	upsmon 	35018 	Poll UPS [pfSense@localhost] failed - Driver not connected 

Thanks.

**Comment by woffko** [comment]
After configuring **SNMP **UPS  I've run into old issue.
ERROR: NUT is enabled, however the service is not running! The SNMP UPS may be unreachable.
But my pfSense is not clean install and I going to try clean install later.

**Comment by woffko** [comment]
Clean Install. Same error.
Manual start from console
[2.3.1-DEVELOPMENT][root@pfSense.localdomain]/root: /usr/local/etc/rc.d/nut onestart
Network UPS Tools - UPS driver controller END-OF-2015Q4-14339-ge69cccf
Can't start /usr/local/libexec/nut/snmp-ups: No such file or directory
/usr/local/etc/rc.d/nut: WARNING: failed precmd routine for nut
You should include SNMP UPS driver into package!

**Comment by Andrew17856** [comment]
Yes,  I noticed the "no such file or directory" error in my setup.  Are all the dependencies there?

**Comment by woffko** [comment]
May be its better to install original freebsd nut port instead of modifying it for pfsense but it's now a little bit complicated and breaking community rules maybe also, I don't know.

**Comment by cbuechler** [comment]
This has been merged. Thanks @sander1988 and @PiBa-NL!

**Comment by sander1988** [comment]
@woffko - You should not start NUT from the command line. Can you check what happens when you start the service through the PFSense webinterface (Status > Services)?

Please also include more details of your NUT settings (all the settings configured on the NUT settings page) so we are able to test the issue you are experiencing.

**Comment by Andrew17856** [comment]
@sander1988 Thanks for your work on this. I can see the package has now reappeared in pfSense (thanks).  I'm having trouble getting it to work with my UPS.  I've posted the issue on the forums here: https://forum.pfsense.org/index.php?topic=111476.0

Any help gratefully received, though I suspect it might be that Blazer driver doesn't support the Megatec MEC0003 protocol, rather than any issue with the port itself.

**Comment by woffko** [comment]
For feedback, my UPS is working with latest official pfSense build without any additional steps required.
Thanks everybody!

**Comment by Andrew17856** [comment]
I do wonder if some of the drivers are missing from the port.  Selecting a number of the drivers from the pull down menu in pfSense gives a "file not found" for the corresponding driver that should be in /usr/local/libexec/nut/ but isn't.

e.g. Selecting the  Mustek UPSs give the following error:
[2.3-RELEASE][admin@pfsense.home]/root: /usr/local/etc/rc.d/nut.sh start
starting megatec
Network UPS Tools - UPS driver controller 2.7.3
Can't start /usr/local/libexec/nut/megatec: No such file or directory
megatec failed to start

Similarly, selecting the Cyber Power UPSs give the following error:

[2.3-RELEASE][admin@pfsense.home]/root: /usr/local/etc/rc.d/nut.sh start
starting cyberpower
Network UPS Tools - UPS driver controller 2.7.3
Can't start /usr/local/libexec/nut/cyberpower: No such file or directory
cyberpower failed to start

**Comment by PiBa-NL** [comment]
Wierd thing is those drivers didn't exist in 2.2 either, where the package was ported from.. Perhaps the whole list of available drivers should be filtered to those drivers that actually exist in /usr/local/libexec/nut/ , and add the few that aren't in the list hoping they will work with the same cable/type parameters..

**Comment by dennypage** [comment]
Hey guys, just wanted to say thank you for your work in getting this into 2.3. Really appreciate it.

**Comment by hansmuff** [comment]
I installed 2.3.2-RELEASE. I had nut configured and running wonderfully on 2.3.1. After upgrading to 2.3.2-RELEASE  (via SSH option 13) pfsense works fine, and all packages except nut work.

When I start nut via Services Status, it does not start. When I use the menu to navigate to Services>NUT, I'll get a 404 message from nginx

I have SSH'd to the box to find that my previous config files all were gone. So I rebuilt them from scratch and ran '/usr/local/etc/rc.d/nut start' and NUT started and seemed to run fine. I still got a 404 on the Services>NUT page though.

So I rebooted pfsense, all the same issues were still there, also something had deleted all my freshly built NUT setup files.. just poof gone, only .sample files left.

I tried uninstall and reinstall of the package, all the same problems.

**Comment by dennypage** [comment]
Please see the information/help link for the package. It looks like a large lower case "i" in the package manager.

https://forum.pfsense.org/index.php?topic=115349.0

**Pull request merged**